src/main/java/org/cardanofoundation/cfexploreraggregator/txcount/processor/AddressTxCountProcessor.java

```diff
@@ -33,52 +43,79 @@ public class AddressTxCountProcessor {
         });
-
     }
 
     private void processTransaction(long slot, String address) {
-        List<AddressTxCountEntity> byAddressOrderBySlotDesc = addressTxCountRepository.findByAddressOrderBySlotDesc(address);
-        if(byAddressOrderBySlotDesc.isEmpty()){
-            AddressTxCountEntity addressTxCountEntity = AddressTxCountEntity.builder()
-                    .address(address)
-                    .txCount(0L)
-                    .slot(slot)
-                    .build();
-            byAddressOrderBySlotDesc.add(addressTxCountEntity);
-        }
-        AddressTxCountEntity addressTxCountEntity = byAddressOrderBySlotDesc.getFirst();
-        addressTxCountEntity.setSlot(slot);
-        addressTxCountEntity.setTxCount(addressTxCountEntity.getTxCount() + 1);
-        addressTxCountRepository.save(addressTxCountEntity);
-
-        // clean up process
-        if(slot - byAddressOrderBySlotDesc.getFirst().getSlot() > 10 ) {
-            addressTxCountRepository.deleteAll(byAddressOrderBySlotDesc);
-        }
+        hashCounts.merge(address, new Tuple<>(slot, 1L), (longLongTuple, longLongTuple2) -> {
+            longLongTuple._1 = Math.max(longLongTuple._1, longLongTuple2._1);
+            longLongTuple._2 += longLongTuple2._2;
+            return longLongTuple;
+        });
     }
 
+    private void flushToDatabase() {
+        try {
+            Map<String, Tuple<Long, Long>> snapshot = new HashMap<>(hashCounts);
+            hashCounts.clear();
+
+            List<AddressTxCountEntity> entities = snapshot.entrySet().stream()
+                    .map(entry -> {
+                        List<AddressTxCountEntity> byAddressOrderByTxCountDesc = addressTxCountRepository.findByAddressOrderBySlotDesc(entry.getKey());
+                        long txCount = 0L;
+                        long slot = 0L;
+                        if(!byAddressOrderByTxCountDesc.isEmpty()) {
+                            txCount = byAddressOrderByTxCountDesc.getFirst().getTxCount();
+                            slot = byAddressOrderByTxCountDesc.getFirst().getSlot();
+                        }
+
+                        return AddressTxCountEntity.builder()
+                                .address(entry.getKey())
+                                .txCount(txCount + entry.getValue()._2)
+                                .slot(Math.max(slot, entry.getValue()._1))
+                                .build();
+                    })
+                    .toList();
+            addressTxCountRepository.saveAll(entities);
+            log.info("Saved {} address tx count entities", entities.size());
+            // getting the max slot from the snapshot and cleaning up the DB
+            snapshot.entrySet().stream().max((o1, o2) -> (int) (o1.getValue()._1 - o2.getValue()._1))
+                    .ifPresent(entry -> addressTxCountRepository.cleanUpDB(entry.getValue()._1 - safeSlotDistance));
```

Comment:
I like this approach, instead of having a job to periodically purge old stuff, as last step of event processing we see if we have in the "update logs" more than one updates that can't be rolled back, and we delete the history up until the first rollback safe event.

This works really well with the `CommitEvent` event.